include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -926,57 +856,191 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
 	const float ay = params_getAlphaV(params);
 	const float ay2 = ay*ay;
 	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data,op);
-	const bool is_bsdf = !op_isBRDF(op);
-	const vec3 refl = params_getReflectance(params);
-
-#ifndef NO_TWOSIDED
-	bool ts_flag = false;
-	handleTwosided_interactionOnly(ts_flag, instr);
-#endif //NO_TWOSIDED
+	const mat2x3 ior2 = matrixCompMult(ior,ior);
+	const bool is_bsdf = !op_isBRDF(op) && !is_coat;
+	const vec3 albedo = params_getReflectance(params);
 
+	float pdf = 1.0;
+	vec3 rem = vec3(1.0);
 	uint ndf = instr_getNDF(instr);
-	irr_glsl_BSDFSample s;
-#ifdef OP_DIFFUSE
-	if (op==OP_DIFFUSE) {
-		s = irr_glsl_oren_nayar_cos_generate(currInteraction, rand.xy, ax2);
-	} else 
-#endif //OP_DIFFUSE
+	irr_glsl_LightSample s;
 
-#ifdef OP_DIFFTRANS
-	if (op==OP_DIFFTRANS) {
-		//TODO take into account full sphere
-		s = irr_glsl_cos_weighted_cos_generate(currInteraction, rand.xy);
+	const vec3 localV = irr_glsl_getTangentSpaceV(currInteraction);
+	const mat3 tangentFrame = irr_glsl_getTangentFrame(currInteraction);
+
+#ifdef OP_DELTATRANS
+	if (op == OP_DELTATRANS)
+	{
+		s = irr_glsl_createLightSample(-precomp.V, -1.0, currInteraction.T, currInteraction.B, currInteraction.isotropic.N);
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+		rem = vec3(1.0);
+		pdf = irr_glsl_FLT_INF;
 	} else
-#endif //OP_DIFFTRANS
+#endif
+#ifdef OP_THINDIELECTRIC
+	if (op == OP_THINDIELECTRIC)
+	{
+		const vec3 luminosityContributionHint = CIE_XYZ_Luma_Y_coeffs;
+		s = irr_glsl_thin_smooth_dielectric_cos_generate(currInteraction, u, ior2[0], luminosityContributionHint);
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+		rem = irr_glsl_thin_smooth_dielectric_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, ior2[0], luminosityContributionHint);
+	} else
+#endif
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+	if (op_isDiffuse(op))
+	{
+		vec3 localL = irr_glsl_projected_hemisphere_generate(u.xy);
+#ifndef NO_BSDF
+		if (is_bsdf)
+		{
+			float dummy;
+			bool flip = irr_glsl_partitionRandVariable(0.5, u.z, dummy);
+			localL = flip ? -localL : localL;
+		}
+#endif
+		s = irr_glsl_createLightSampleTangentSpace(localV, localL, tangentFrame);
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
 
-//TODO need to distinguish BSDFs and BRDFs
-#ifdef NDF_GGX
-	if (ndf == NDF_GGX) {
-		s = irr_glsl_ggx_cos_generate(currInteraction, rand.xy, ax, ay);
+		rem *= albedo*irr_glsl_oren_nayar_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, ax2);
+		pdf *= is_bsdf ? 0.5 : 1.0;
 	} else
+#endif
+#if defined(OP_CONDUCTOR) || defined(OP_DIELECTRIC)
+	if (op_hasSpecular(op))
+	{
+		const float TdotV2 = currInteraction.TdotV * currInteraction.TdotV;
+		const float BdotV2 = currInteraction.BdotV * currInteraction.BdotV;
+		const float NdotV2 = currInteraction.isotropic.NdotV_squared;
```

Comment:
similar comment to https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r528886851